crates/ty_python_semantic/resources/mdtest/type_of/generics.md

```diff
@@ -0,0 +1,163 @@
+# `type[T]`
+
+`type[T]` with a type variable represents the class objects of `T`.
+
+```toml
+[environment]
+python-version = "3.13"
+```
+
+## Basic
+
+The meta-type of a typevar is `type[T]`.
+
+```py
+from typing import reveal_type
+
+def _[T](x: T):
+    reveal_type(type(x))  # revealed: type[T@_]
+```
+
+`type[T]` with an unbounded type variable represents any subclass of `object`.
+
+```py
+def unbounded[T](x: type[T]) -> T:
+    reveal_type(x)  # revealed: type[T@unbounded]
+    reveal_type(x.__repr__)  # revealed: def __repr__(self) -> str
+    reveal_type(x.__init__)  # revealed: def __init__(self) -> None
+    reveal_type(x())  # revealed: Any
+
+    return x()
+```
+
+`type[T]` with an upper bound of `T: A` represents any subclass of `A`.
+
+```py
+class A:
+    x: str
+
+    def __init__(self, value: str): ...
+
+class B(A): ...
+class C: ...
+
+def upper_bound[T: A](x: type[T]) -> T:
+    reveal_type(x)  # revealed: type[T@upper_bound]
+    reveal_type(x("hello"))  # revealed: T@upper_bound
+
+    return x("hello")
+
+reveal_type(upper_bound(A))  # revealed: A
+reveal_type(upper_bound(B))  # revealed: B
+
+# error: [invalid-argument-type] "Argument to function `upper_bound` is incorrect: Argument type `C` does not satisfy upper bound `A` of type variable `T`"
+upper_bound(C)
+```
+
+`type[T]` with a constraints `T: (A, B)` represents exactly the class object `A`, or exactly `B`:
+
+```py
+def constrained[T: (int, str)](x: type[T]) -> T:
+    reveal_type(x)  # revealed: type[T@constrained]
+
+    # TODO: This should reveal `T`.
+    reveal_type(x("hello"))  # revealed: int | Literal["hello"]
```

Comment:
This doesn't currently work because we can only treat `type[T]` as a class constructor if `T` has an upper bound of a single concrete class. If there are multiple constraints, we fall back to the call binding logic, which splits `type[T]` into the union of its constraints.

I think we have a general problem of not unifying operations on constrained type variables very well (e.g. [this failing test](https://github.com/astral-sh/ruff/blob/40886963d1c01bd00f27f2b67913504a52d2d85f/crates/ty_python_semantic/resources/mdtest/generics/pep695/functions.md?plain=1#L250)), so I'm not sure this issue can be addressed by this PR.